packages/web/hooks/swap/use-send-swap-tx.ts

```diff
@@ -0,0 +1,192 @@
+  }
+
+  let coinAmount = inAmountInput.amount;
+
+  // check that the amount left is enough to cover fees
+  const amountLeft = inAmountInput.balance.sub(inAmountInput.amount).toDec();
+
+  let amountAdjustedForFees = false;
+  // update a threshhold for calculating furhter the estimate, maybe from preferences max gas fees
+  let local_networkFee =
+    featureFlags.swapToolSimulateFee && networkFee ? networkFee : undefined;
+  // if the amount left in balance is below AMOUNT_LEFT_THRESHOLD, subtract the gas estimates from the amount to transfer, and set amountAdjustedForFees flag.
+  if (amountLeft.lt(AMOUNT_LEFT_THRESHOLD)) {
+    local_networkFee =
+      local_networkFee || (await runEstimateTxFeesOnce(messages));
+    const adjustment = local_networkFee!.gasAmount;
+    coinAmount = coinAmount.sub(adjustment);
+    amountAdjustedForFees = true;
+  }
+
+  const fee = local_networkFee && {
+    preferNoSetFee: true,
+    fee: {
+      gas: local_networkFee.gasLimit,
+      amount: local_networkFee.amount,
+    },
+  };
+
+  const fromToken = fromAsset?.coinDenom;
+  const toToken = toAsset?.coinDenom;
+  const tokenAmount = Number(coinAmount);
+  const isMultiHop =
+    quote?.split.some(({ pools }) => pools.length !== 1) ?? false;
+  const isMultiRoute = (quote?.split.length ?? 0) > 1;
+  const quoteTimeMilliseconds = quote?.timeMs;
+  const router = quote?.name;
+  const amountFiatValue = quote?.amountFiatValue;
+
+  return {
+    fee,
+    coinAmount,
+    routes,
+    tokenIn,
+    tokenOutMinAmount,
+    fromToken,
+    tokenAmount,
+    toToken,
+    isMultiHop,
+    isMultiRoute,
+    quoteTimeMilliseconds,
+    router,
+    amountFiatValue,
+    amountAdjuestedForFees: amountAdjustedForFees,
+  };
+};
+
+/**
+ * A React Query mutation sends a swap transaction.
+ */
+export const useSendSwapTxMutation = () => {
+  const { logEvent } = useAmplitudeAnalytics();
+  return useMutation({
+    mutationFn: async ({
+      fee,
+      coinAmount,
+      slippage,
+      page,
+      amountFiatValue,
+      routes,
+      tokenIn,
+      tokenOutMinAmount,
+      osmosis,
+      ...baseEvent
+    }: SwapTransactionData) => {
+      logEvent([EventName.Swap.swapStarted, baseEvent]);
+
+      let result: "multiroute" | "multihop" | "exact-in" | undefined =
+        undefined;
+      if (routes.length === 1) {
+        const { pools } = routes[0];
+        osmosis.sendSwapExactAmountInMsg(
+          pools,
+          tokenIn,
+          tokenOutMinAmount,
+          undefined,
+          fee,
+          () => {
+            result = pools.length === 1 ? "exact-in" : "multihop";
+          }
+        );
+      } else if (routes.length > 1) {
+        osmosis.sendSplitRouteSwapExactAmountInMsg(
+          routes,
+          tokenIn,
+          tokenOutMinAmount,
+          undefined,
+          fee,
+          () => {
+            result = "multiroute";
+          }
+        );
+      } else {
+        throw new Error("No routes given");
+      }
```

Comment:
We should encapsulate the send operations within a [promise](https://github.com/osmosis-labs/osmosis-frontend/blob/dbdf9c943695fa339682e3a18f12373a77c81287/packages/web/hooks/mutations/one-click-trading/use-create-one-click-trading-session.tsx#L377-L396) and use await to ensure we prevent memory leaks and wait for the operation to complete.

Reply:
3e813134be3d12cb3854dbbfaff381917b369f64